# Homework 04

In this assignment, you will practice wrangling data through different formats, with missing data, and working with text.

You may or may not use `for` loops depending on the questions. **Please see the question instructions carefully to determine if `for` loops can be used.** Generally, you should never be iterating over the rows of a DataFrame using `for` loops. If you do, you will lose points.

## Part 1: Text wrangling and regular expressions
In this part, we will work with the citation file exported from the [Nature Review Article](https://www.nature.com/articles/s41586-020-2649-2) *Array Programming with NumPy*. Below we read the file into the Python variable `cite` and print the result for you to preview.

**It is fine to use `for` loops in this part since it involves parsing code before putting them into a `DataFrame`.**

In [1]:
# Run but do not modify this code
with open("numpy_nature.txt", encoding="utf-8") as f:
    cite = f.read()

print(cite)


TY  - JOUR
AU  - Harris, Charles R.
AU  - Millman, K. Jarrod
AU  - van der Walt, Stéfan J.
AU  - Gommers, Ralf
AU  - Virtanen, Pauli
AU  - Cournapeau, David
AU  - Wieser, Eric
AU  - Taylor, Julian
AU  - Berg, Sebastian
AU  - Smith, Nathaniel J.
AU  - Kern, Robert
AU  - Picus, Matti
AU  - Hoyer, Stephan
AU  - van Kerkwijk, Marten H.
AU  - Brett, Matthew
AU  - Haldane, Allan
AU  - del Río, Jaime Fernández
AU  - Wiebe, Mark
AU  - Peterson, Pearu
AU  - Gérard-Marchant, Pierre
AU  - Sheppard, Kevin
AU  - Reddy, Tyler
AU  - Weckesser, Warren
AU  - Abbasi, Hameer
AU  - Gohlke, Christoph
AU  - Oliphant, Travis E.
PY  - 2020
DA  - 2020/09/01
TI  - Array programming with NumPy
JO  - Nature
SP  - 357
EP  - 362
VL  - 585
IS  - 7825
AB  - Array programming provides a powerful, compact and expressive syntax for accessing, manipulating and operating on data in vectors, matrices and higher-dimensional arrays. NumPy is the primary array programming library for the Python language. It has an essential r

### Question 1 (4 points)
There are several authors, each recorded on a separate line beginning with `AU`. In the variable `q1` put a Python list of all of the author names formatted as in the file but without the extra characters and whitespace (i.e., without the `AU  - ` or the newline `\n` characters). Your list should be of the form `['Harris, Charles R.', 'Millman, K. Jarrod', ..., 'Oliphant, Travis E.']`. When you are finished, print the resulting list. 

In [2]:
# Put your code to answer the question here, and store the answer as the following variable
# Please keep author names in the original order

AU_lst = []

for item in cite.split("\n"):
    if "AU  - " in item:
        AU_lst.append(item[6:])

q1 = AU_lst


print(q1)

['Harris, Charles R.', 'Millman, K. Jarrod', 'van der Walt, Stéfan J.', 'Gommers, Ralf', 'Virtanen, Pauli', 'Cournapeau, David', 'Wieser, Eric', 'Taylor, Julian', 'Berg, Sebastian', 'Smith, Nathaniel J.', 'Kern, Robert', 'Picus, Matti', 'Hoyer, Stephan', 'van Kerkwijk, Marten H.', 'Brett, Matthew', 'Haldane, Allan', 'del Río, Jaime Fernández', 'Wiebe, Mark', 'Peterson, Pearu', 'Gérard-Marchant, Pierre', 'Sheppard, Kevin', 'Reddy, Tyler', 'Weckesser, Warren', 'Abbasi, Hameer', 'Gohlke, Christoph', 'Oliphant, Travis E.']


### Question 2 (4 points)
Create a Pandas DataFrame that contains three columns: one for first names, one for middle names, and one for last names for all of the authors. Use the column names of the example table below. Keep the same order as the original text file and use the default primary index (the row labels) of 0, 1, 2, etc. as shown below. You are welcome to use the results of the prior question to answer this problem.

|      | first      | middle     | last         |
| ---- | ---------- | ---------- | ------------ |
| 0	   | Charles    | R.         | Harris       |
| 1	   | K.	        | Jarrod     | Millman      |
| 2	   | Stéfan     | J.         | van der Walt |
| 3    | Ralf       |            | Gommers      |
| 4	   | Pauli      |            | Virtanen     |   

Note that some authors do not have any middle names, in which case you should set the middle name as an empty string `''`.

In [3]:
# Run but do not modify this code
import pandas as pd

In [4]:
# Put your code to answer the question here, and store the answer as the following variable
# Please keep author names in the original order

fname = []
mname = []
lname = []

for name in AU_lst:
    parts = name.split(", ")
    lname.append(parts[0])

    first_half = parts[1].split(" ")
    
    fname.append(first_half[0])
    
    if len(first_half) > 1:
        mname.append(" ".join(first_half[1:]))
    else:
        mname.append("")

q2 = pd.DataFrame({"first": fname, "middle": mname, "last": lname})

### Question 3 (12 points)
Below we extract the abstract from the citation and store it in a string variable `abstract`. Write regular expressions to answer the following questions about the abstract.

1. In `q3_1` put the starting index of everywhere `NumPy` appears in the abstract (i.e., the index of the `N` wherever `NumPy` occurs in the `abstract` string). This should be case sensitive.
2. In `q3_2` put all of the capitalized words in `abstract`, including words with extra capitalized letters like `NumPy`, `NumPy-like`, and `API`. Note that we keep `-` since this is part of a word, but do not keep the `()` that are outside `API`. 
3. In `q3_3` put all of the words that immediately follow `NumPy`, but do not include the word `NumPy` itself. For the one occurrence it is hyphenated `NumPy-like`, use `-like`.

All three of your answers should be of type `list` and if something appears multiple times in the abstract it should also appear multiple times in your answer

In [5]:
# Run but do not modify this code
import re
abstract_query = re.compile(r"AB  - (.+)")
abstract = re.search(abstract_query, cite).group(1)
print(abstract)

Array programming provides a powerful, compact and expressive syntax for accessing, manipulating and operating on data in vectors, matrices and higher-dimensional arrays. NumPy is the primary array programming library for the Python language. It has an essential role in research analysis pipelines in fields as diverse as physics, chemistry, astronomy, geoscience, biology, psychology, materials science, engineering, finance and economics. For example, in astronomy, NumPy was an important part of the software stack used in the discovery of gravitational waves1 and in the first imaging of a black hole2. Here we review how a few fundamental array concepts lead to a simple and powerful programming paradigm for organizing, exploring and analysing scientific data. NumPy is the foundation upon which the scientific Python ecosystem is constructed. It is so pervasive that several projects, targeting audiences with specialized needs, have developed their own NumPy-like interfaces and array object

In [6]:
# Put your code to answer the question here, and store the answer as the following variable
# Please allow duplication in your list

q3_1 = [m.start() for m in re.finditer(r"NumPy", abstract)]

q3_2 = re.findall(r"\b[A-Z][\w-]*", abstract)

words_q3_3 = re.findall(r"(?<=NumPy)(-\w+|\s\w+)", abstract)
q3_3 = [m.strip() for m in words_q3_3]

# Run but do not modify this code
print('q3_1:', q3_1)
print('q3_2:', q3_2)
print('q3_3:', q3_3)

q3_1: [171, 469, 768, 962, 1051]
q3_2: ['Array', 'NumPy', 'Python', 'It', 'For', 'NumPy', 'Here', 'NumPy', 'Python', 'It', 'NumPy-like', 'Owing', 'NumPy', 'API']
q3_3: ['is', 'was', 'is', '-like', 'increasingly']


## Part 2: Cleaning up more system logs CSV

In this part, we work with a piece of messy tabular data in the form of a poorly formatted csv file containing data about programs running on computer systems. It contains all of the data about system time and memory from Worked Example 4, but also includes new information about user ids and machine ids, and some data are missing in every column. (The user ids are made up and do not correspond to any real individuals).

**Do not** use `for` loops, unless it is guaranteed to iterate less than 50 times. The manual points are partially for checking for this. You are also likely to not pass some tests because iterating over the entire `DataFrame` sometimes causes issues with formatting.

### Question 4 (12 points, 2 manual points)

Below, we import the dataset using the Pandas `read_csv` function that creates a dataframe. Run the code; it will preview the first five rows.

In [7]:
# Run but do not modify this code
import pandas as pd
import numpy as np
sys_df = pd.read_csv("more_monitor.csv")
sys_df.head()

,System Time second,System Memory GB,System Memory MB,System User ID,System Machine ID
0,?,?,?,User ID: yw22,Machine ID: Carrot
1,System Time: 40 second,System Mem: 3 Gb,382 Mb,?,?
2,?,System Mem: 2 Gb,271 Mb,User ID: tp7,Machine ID: Asparagus
3,System Time: 31 second,System Mem: 3 Gb,493 Mb,?,Machine ID: Eggplant
4,System Time: 37 second,System Mem: 3 Gb,411 Mb,?,Machine ID: Asparagus


There are several formatting issues with the default import. Address the following.

1. The `System User ID` and `System Machine ID` contain String data with the redundant information `User ID: ` and `Machine ID: ` in every row that has data. Remove these prefixes so that the columns only contain the user ids and machine ids themselves. Leave the `?` cells as is. For example, the first row should just have `yw22` in the `System User ID` column and `Carrot` in the `System Machine ID` column, while the second row should keep `?` for both columns.

2. The first three columns for `System Time second`, `System Memory GB` and `System Memory MB` contain numerical data but are currently formatted as strings with redundant prefix information repeating the column label and missing data represented as the string `?` instead of the Numpy `np.nan` sentinel value. Fix this so that each value in the first three columns is either a single numerical value or NaN (note, you should use the actual `np.nan` sentinal value, not just the String with the characters `n`, `a`, and `n`). For example, when you are done, the first three columns of the first row should all have NaN values, the second row should be `40`, `3`, and `382`, and so on. Note that the rows at index `400` and on have System Time recorded in minutes instead of seconds, be sure to convert these to seconds by mulitplying by 60.

3. Currently the System Memory is split accross two columns, one for the GB and one for the MB. For example, the total memory of the first program is 3 GB and 414 MB. Instead, represent the full system memory in the `System Memory GB` column, and get rid of the `System Memory MB` column. To do so, you need to convert the values in the MB column to GB (1 MB is 0.001 GB) and add that to the GB column, then use the [`drop` method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html). If a value is missing (and should now be a NaN), your result should maintain its value as missing/unknown.

When you are finished, `sys_df` should have the above issues corrected. Run both of the cells with `sys_df.head()` and `sys_df.tail()` to show the first and last few rows of your dataframe. Create a variable `q4` and have it point to this DataFrame.

In [8]:
# Put your code to answer the question here, and store sys_df in q4 for full credit

user_mask = sys_df['System User ID'].notna() & (sys_df['System User ID'] != '?')
sys_df.loc[user_mask, 'System User ID'] = (
    sys_df.loc[user_mask, 'System User ID']
      .str.replace(r'^\s*User ID:\s*', '', regex=True)
      .str.strip()
)

machine_mask = sys_df['System Machine ID'].notna() & (sys_df['System Machine ID'] != '?')
sys_df.loc[machine_mask, 'System Machine ID'] = (
    sys_df.loc[machine_mask, 'System Machine ID']
      .str.replace(r'^\s*Machine ID:\s*', '', regex=True)
      .str.strip()
)

time_s = sys_df['System Time second'].astype(str)

time_num = pd.to_numeric(time_s.str.extract(r'(\d+(?:\.\d+)?)', expand=False), errors='coerce')

text_min_mask = time_s.str.contains(r'\bmin(?:ute)?s?\b', case=False, na=False)
index_min_mask = sys_df.index.to_series().ge(400)
minute_mask = text_min_mask | index_min_mask

time_num = time_num.astype('float')
time_num.loc[minute_mask] = time_num.loc[minute_mask] * 60

sys_df['System Time second'] = time_num

gb_s = sys_df['System Memory GB'].astype(str)
mb_s = sys_df['System Memory MB'].astype(str)

gb_num = pd.to_numeric(gb_s.str.extract(r'(\d+(?:\.\d+)?)', expand=False), errors='coerce')
mb_num = pd.to_numeric(mb_s.str.extract(r'(\d+(?:\.\d+)?)', expand=False), errors='coerce')

total_gb = gb_num + (mb_num * 0.001)

sys_df['System Memory GB'] = total_gb

sys_df = sys_df.drop(columns=['System Memory MB'])

q4 = sys_df

# Run but do not modify this code
print(q4.head())

# Run but do not modify this code
print(q4.tail())

   System Time second  System Memory GB System User ID System Machine ID
0                 NaN               NaN           yw22            Carrot
1                40.0             3.382              ?                 ?
2                 NaN             2.271            tp7         Asparagus
3                31.0             3.493              ?          Eggplant
4                37.0             3.411              ?         Asparagus
     System Time second  System Memory GB System User ID System Machine ID
995              1860.0             2.258            zm3         Asparagus
996              1860.0             3.403            bk4                 ?
997              2160.0             3.350           yw22            Carrot
998              2640.0             3.366           yw22         Asparagus
999              1200.0             3.490            bk4         Asparagus


<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

### Question 5 (12 points, 2 manual points)
The `sys_df` dataframe from question 4 should now be a little easier to read and use. Answer the following questions about `sys_df`.

1. How many rows are missing data (is a `?`) in the `System Machine ID` column? Put your answer in `q5_1`.
2. What is the average value of `System Memory GB` among the rows that are missing data (have a `?`) in the `System User ID` column?  Put your answer in `q5_2`.
3. How many rows are missing data in both the `System Time second` and `System Memory GB` columns? Put your answer in `q5_3`.

**Hint**: If your code mutates the dataframe, we recommend you use `.copy()` first so that Q4 is graded independent of any changes you make to the dataframe object. 

In [9]:
# Put your code to answer the question here, and store answers as the following variables

df = q4.copy()

df['System Time second'] = pd.to_numeric(df['System Time second'], errors='coerce')
df['System Memory GB'] = pd.to_numeric(df['System Memory GB'], errors='coerce')

q5_1 = int(((df['System Machine ID'] == '?') | (df['System Machine ID'].isna())).sum())

mask_user_missing = (df['System User ID'] == '?') | (df['System User ID'].isna())
q5_2 = df.loc[mask_user_missing, 'System Memory GB'].mean()

q5_3 = int((df['System Time second'].isna() & df['System Memory GB'].isna()).sum())


# Run but do not modify this code
print('q5_1:', q5_1)
print('q5_2:', q5_2)
print('q5_3:', q5_3)

q5_1: 196
q5_2: 2.909991596638655
q5_3: 69


<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## Part 3: Wrangling FDA JSON Dataset 
In this part we work with a messy JSON dataset containing information about several drugs labels.

### Question 6 (12 points)
Below we import the `FDADrugLabel.json` file into the `labels` variable. This is the same dataset as the worked example. The resulting Python object is somewhat messy; we encourage you to explore the data before answering the questions.

**You may use `for` loops for this question.**

In [10]:
# Run but do not modify this code
import json
with open("FDADrugLabel.json", encoding="utf-8") as f:
    labels = json.load(f)

In [11]:
# Use this cell to explore (and you are welcome to add more cells as needed)
len(labels[0])

15

Answer the following questions.

1. In `q6_1` put the mean number of key/value (or name/value) pairs in the top level dictionary for each drug. For example, the first element in `labels` has 15 pairs in its top level dictionary.

2. In `q6_2` put the list of all of the `manufacturer_names` without any other information. `manufacturer_names` are not a top level key/name, you will need to search for where they are located and how to extract them.

3. In `q6_3` put how many drugs contain the string `child` anywhere in their `warnings` (including as part of larger strings like `children`). Note: `warnings` is a top level key/name.

In [12]:
# Put your code to answer the question here, and store answers as the following variables
for drug in labels:
    total_keys = 0
    for item in labels:
        total_keys += len(item.keys())

q6_1 = total_keys / len(labels)

q6_2 = [item["openfda"]["manufacturer_name"] for item in labels]

q6_3 = len([item for item in labels if "warnings" in item and "child" in "".join(item["warnings"])])

# Run but do not modify this code
print('q6_1:', q6_1)
print('q6_2:', q6_2)
print('q6_3:', q6_3)

q6_1: 19.25
q6_2: ['Nature and Health Beauty Co., Ltd.', 'Silver Star Brands', 'Johnson & Johnson Consumer Inc.', 'Proficient Rx LP', 'Energique, Inc.', 'Energique, Inc.', 'Amerisource Bergen', 'Seroyal USA', 'Proficient Rx LP', 'MODECOS CO., LTD.', 'United Exchange Corp.', 'King Bio Inc.', 'Aurobindo Pharma Limited', 'Energique, Inc.', 'SMART SENSE (Kmart)', 'Mentor Lab', 'BCM Ltd', 'Pearl World INC.', 'Dolgencorp, Inc. (DOLLAR GENERAL & REXALL)', 'Medline Industries, Inc.']
q6_3: 13


### Question 7 (8 points, 2 manual points)

Below we use the Pandas `read_json` function to import the `FDADrugLabel.json` file into a Pandas dataframe `label_df`.

In [13]:
# Run but do not modify this code
import pandas as pd
label_df = pd.read_json("FDADrugLabel.json")
label_df.head()

,package_label_principal_display_panel,active_ingredient,warnings,inactive_ingredient,drug_interactions,effective_time,openfda,keep_out_of_reach_of_children,spl_product_data_elements,set_id,...,ask_doctor,do_not_use,recent_major_changes,boxed_warning,indications_and_usage_table,precautions,labor_and_delivery,general_precautions,dosage_and_administration_table,ask_doctor_or_pharmacist
0,[label],titanium dioxide,[1. Do not use in the following cases(Eczema a...,"butylene glycol, water, etc",,20160321,{'manufacturer_name': 'Nature and Health Beaut...,[keep out of reach of the children],"[Nature and Health Beauty (Moisturizing Cream,...",2ea9b21c-18e6-0a8a-e054-00144ff8d46c,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,[BrightSpark Oral Spray],HPUS Active Ingredients Each dose contains equ...,"[Warnings If symptoms persist or worsen, consu...","Citric acid, potassium sorbate, purified water.",NaN,20160323,"{'manufacturer_name': 'Silver Star Brands', 'u...",[Keep this and all medication out of reach of ...,"[BrightSpark Arg nit, Arsenicum iod, Hyoscyamu...",2eb93d7c-330f-4b93-e054-00144ff88e88,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,[PRINCIPAL DISPLAY PANEL - 106 g Packet Carton...,Active ingredient Colloidal oatmeal 43%,[Warnings For external use only. When using th...,"Inactive ingredients mineral oil, calcium sili...",NaN,20160315,{'manufacturer_name': 'Johnson & Johnson Consu...,"[Keep out of reach of children. If swallowed, ...",[Aveeno Baby Eczema Therapy Soothing Bath Trea...,7f689b72-ccfb-4ae0-9747-1490593998ca,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,[PACKAGE LABEL-PRINCIPAL DISPLAY PANEL NDC 631...,NaN,NaN,NaN,[7 DRUG INTERACTIONS In vitro data indicate th...,20160301,"{'manufacturer_name': 'Proficient Rx LP', 'uni...",NaN,[levocetirizine dihydrochloride levocetirizine...,b84ccb27-9c52-43b9-9a46-8f52a571045e,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,[PACKAGE LABEL DISPLAY: ENERGIQUE SINCE 1987 H...,"ACTIVE INGREDIENTS: Angelica Archangelica 3X, ...","[WARNINGS: If pregnant or breastfeeding, ask a...","INACTIVE INGREDIENTS: Demineralized water, 20%...",NaN,20160322,"{'manufacturer_name': 'Energique, Inc.', 'unii...",[KEEP OUT OF REACH OF CHILDREN: Keep out of re...,"[Ovapar Angelica Archangelica, Caulophyllum Th...",13ecc450-0ae3-46d3-8cc7-915c74c3361a,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Use the dataframe to answer the following questions. You may use `for` loops for these questions only after selecting for the subset of data you are looking for. Do not iterate through all of the rows in the dataframe.

1. In `q7_1` put a python list of the index numbers of the drugs with a value (not null) for the `ask_doctor` column. 

1. In `q7_2` put a dictionary where the key is the `generic_name` of the drug and the value is the value in the `nursing_mothers` column. Only include the 3 drugs with an actual value for the `nursing_mothers` column. The `generic_name` can be found in the dictionary in the `openfda` column.

In [14]:
# Put your code to answer the question here, and store answers as the following variables

q7_1 = label_df[label_df["ask_doctor"].notna()].index.tolist()

nm_df = label_df[label_df["nursing_mothers"].notna()].head(3)

q7_2 = {}
for idx, row in nm_df.iterrows():
    gen = row.get("openfda", {}).get("generic_name", None)
    if isinstance(gen, list):
        generic_name = gen[0]
    else:
        generic_name = gen

    if generic_name is None:
        generic_name = "UNKNOWN"
    else:
        generic_name = str(generic_name).strip()

    nm_val = row["nursing_mothers"]
    if isinstance(nm_val, list):
        nm_val = nm_val[0]
    else:
        nm_val = str(nm_val)

    q7_2[generic_name] = nm_val

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

### Question 8 (6 points, 3 manual points)

Do the following. You may **not** use `for` loops for this question:
1. Use a regular expression(s) to create at least one new column in `label_df` by extracting information from an existing column with data of type string. Put the column's name in `new_col_name`.
1. Count the number of NaNs in **the column you extracted from** and put it in `nan_count_org`.
1. Count the number of NaNs in **your new column** and put it in `nan_count_new`.

Note: If you create more than one new column or use more than one existing column in `label_df`, choose just one to fill in `nan_count_org` and `nan_count_new`.

In the written part, provide an explanation of:
1. What the code you wrote in the code cell does.
1. Whether there are NaNs in the original and new column you created for `label_df`. Provide a possible reason why the NaNs are/are not there.
1. Why do the original and new columns have the same or not the same number of NaNs.

**Grading of this question**: It will be graded using an ESNU rubric as follows:
- Exemplary - Work that meets all requirements and displays full mastery of all learning goals for this question. And the code and write-up are clean and easy to understand.
- Satisfactory - Work that meets all requirements and displays at least partial mastery of all learning goals as well as full mastery of core learning goals. In other words, it is not an E, but it fulfills the requirements, including (but not necessarily only) the code matches what the write-up claims it is.
- Not Yet - Work that does not meet some requirements and/or displays developing or incomplete mastery of at least some learning goals. For example, the code and write-up do not match, or the write-up states something incorrectly.
- Unasessable - Work that is missing, does not demonstrate meaningful effort, or does not provide enough evidence to determine a level of mastery.

In [15]:
orig_col = "nursing_mothers"

new_col_name = "nm_excreted_in_milk"

label_df[new_col_name] = label_df[orig_col].str.extract(r"(excreted in milk)", flags=re.IGNORECASE)[0]

nan_count_org = label_df[orig_col].isna().sum()
nan_count_new = label_df[new_col_name].isna().sum()


# Leave these here to help with grading.
print('New column:', new_col_name)
print('Head of new column:')
print(label_df[new_col_name].head())
print('Count of NaNs in original column:', nan_count_org)
print('Count of NaNs in new column:', nan_count_new)

New column: nm_excreted_in_milk
Head of new column:
0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: nm_excreted_in_milk, dtype: object
Count of NaNs in original column: 17
Count of NaNs in new column: 20


### Answer 8: Written portion

1. I selected the column `nursing_mothers` and created a new column named `nm_excreted_in_milk` by looking for the phrases "excreted in milk" or "excreted in human milk" .The `(?:human )?` makes "human " optional and the `flags=re.IGNORECASE` makes the match case-insensitive. When the phrase is found, it is stored in the new column and when it is not found, pandas assigns a NaN value.

2. There are NaN values in both the `nursing_mothers` column and the `nm_excreted_in_milk` column. This is because not every drug label includes a “Nursing Mothers” section and in the new column, NaNs appear because either the original value was already NaN, or the text did not contain the specific phrase.

3. The number of NaNs in the new column is not the same as the number of NaNs in the original column. This is because every NaN in the original column will also result in a NaN in the new column, however regarding the new column, some rows are not NaN in `nursing_mothers` but contain the specific phrase, so the regex finds no match and returns NaN for those rows. This causes the new column to have more NaNs than the original.

<!-- END QUESTION -->

## AI Disclosure

Use the [Artificial Intelligence Disclosure (AID) Framework](https://crln.acrl.org/index.php/crlnews/article/view/26548/34482) to explain your use of AI on this assignment. Other headings you can use include:
- *Debugging*: Using AI to help you fix your code so that it works. You should state how you used it for this purpose.

Here are some examples:

*Artificial Intelligence Tools*: ChatGPT v5 via chatgpt.com. *Conceptualization*: I gave chatgpt.com the election data set and asked it for ideas on interesting statistics I could get from the data. *Methodology*: I asked it for help on how to write the code to get the statistic I chose, but I wrote the code myself. *Writing—Review & Editing*: I wrote out my explanation for what the statistic meant, then gave that text and the rubric to chatgpt and asked it to give me feedback on how to update the explanation to conform to the rubric.

*Artificial Intelligence Tools*: ChatGPT v4o via  DukeGPT. *Information Collection*: DukeGPT was used to find the function needed to get the index value of the maximum value of a Series and the syntax needed to filter rows in a pandas dataframe using multiple columns. *Debugging*: DukeGPT was used to help me find a bug in my code for Q1 where I copied in the code and error, stated what the code should do, and asked for help.

**Solution:**

*Artificial Intelligence Tools*: ChatGPT v5 via chatgpt.com. *Conceptualization*: I gave chatgpt.com the `FDADrugLabel.json` file and asked it for ideas on a new column I could get from the data. *Debugging*: ChatGPT was used to help me find a bug in my code for Q4 where I copied in the code and error, stated what the code should do, and asked for help. *Writing—Review & Editing*: I wrote out my explanation for what my code meant and my data, then gave that text and the rubric to chatgpt and asked it to give me feedback on how to update the explanation to conform to the rubric.

## Submitting

You should make sure any code that you write to answer the questions is included in this notebook. You are **required** to go to the Kernel option and choose **"Restart & Run All"**  before submission. Double check that your entire notebook runs correctly and generates the expected output. Finally, make sure to save your work (timestamp at the top tells you the last checkpoint and whether there are unsaved changes). When you finish, submit your assignment at [Gradescope](http://gradescope.com/ "‌"). **Submissions not prepared correctly as above will lose points.**